packages/protocol-sdk/src/1155-create-helper.ts

```diff
@@ -0,0 +1,319 @@
+  publicClient: PublicClient,
+  contract: ContractType,
+  account: Address,
+) {
+  let contractAddress;
+  let contractExists = false;
+  if (typeof contract !== "string") {
+    contractAddress = await publicClient.readContract({
+      abi: zoraCreator1155FactoryImplABI,
+      // Since this address is deterministic we can hardcode a chain id safely here.
+      address: zoraCreator1155FactoryImplAddress[999],
+      functionName: "deterministicContractAddress",
+      args: [
+        account,
+        contract.uri,
+        contract.name,
+        contract.defaultAdmin || account,
+      ],
+    });
+    try {
+      await publicClient.readContract({
+        abi: zoraCreator1155ImplABI,
+        address: contractAddress,
+        functionName: "contractVersion",
+      });
+      contractExists = true;
+    } catch (e: any) {
+      // This logic branch is hit if the contract doesn't exist
+      //  falling back to contractExists to false.
+    }
+    return { contractAddress, contractExists };
+  }
+
+  return {
+    contractExists: true,
+    contractAddress: contract,
+  };
+}
+
+export async function createNew1155Token({
+  publicClient,
+  contract,
+  tokenMetadataURI,
+  mintToCreatorCount = 1,
+  salesConfig = {},
+  maxSupply,
+  account,
+  royaltySettings,
+  getAdditionalSetupActions,
+}: {
+  publicClient: PublicClient;
+  account: Address;
+  maxSupply?: bigint | number;
+  royaltySettings?: RoyaltySettingsType;
+  royaltyBPS?: number;
+  contract: ContractType;
+  tokenMetadataURI: string;
+  mintToCreatorCount?: bigint | number;
+  salesConfig?: SalesConfigParamsType;
+  getAdditionalSetupActions?: (args: {
+    tokenId: bigint;
+    contractAddress: Address;
+  }) => Hex[];
+}) {
+  const { contractExists, contractAddress } = await getContractExists(
+    publicClient,
+    contract,
+    account,
+  );
+
+  let nextTokenId = 1n;
+
+  if (contractExists) {
+    nextTokenId = await publicClient.readContract({
+      abi: zoraCreator1155ImplABI,
+      functionName: "nextTokenId",
+      address: contractAddress,
+    });
+  }
+
+  const fixedPriceMinterAddress = await publicClient.readContract({
+    abi: zoraCreator1155FactoryImplABI,
+    address: zoraCreator1155FactoryImplAddress[999],
+    functionName: "fixedPriceMinter",
+  });
+
+  let tokenSetupActions = create1155TokenSetupArgs({
+    tokenMetadataURI,
+    nextTokenId,
+    salesConfig,
+    maxSupply,
+    fixedPriceMinterAddress,
+    account,
+    mintToCreatorCount,
+    royaltySettings,
+  });
+  if (getAdditionalSetupActions) {
+    tokenSetupActions = [
+      ...getAdditionalSetupActions({ tokenId: nextTokenId, contractAddress }),
+      ...tokenSetupActions,
+    ];
+  }
+
+  if (!contractAddress && typeof contract === "string") {
+    throw new Error("Invariant: contract cannot be missing and an address");
+  }
+  if (!contractExists && typeof contract !== "string") {
+    const { request } = await publicClient.simulateContract({
```

Comment:
this function is very useful, but it is doing a bit too much, reflected in the long length; id actually break this down into a less opinionated function that returns params that you can submit to viem or wagmi or even ethers.js

Like:

`makePrepareNewToken` - this creates a resulting config object that, and can be plugged into viem or wagmi hooks:

```ts
// makePrepareNewToken is sdk function
const prepareCreateNewTokenParams = makePrepareNewToken({
  maxSupply,
  account,
  royaltySettings,
  getAdditionalSetupActions,
  ...etc
});

// with viem
const { request } = await publicClient.simulateContract(prepareCreateNewTokenParams):

// with wagmi
const { config, error } = usePrepareContractWrite(prepareCreateNewTokenParams);
```

Reply:
Ran into type issues – waiting to chat about a fix.